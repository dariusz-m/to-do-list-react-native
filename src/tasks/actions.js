export const GET_TASKS = 'GET_TASKS';
export const TASKS_HAVE_BEEN_LOADED = 'TASKS_HAVE_BEEN_LOADED';
export const ADD_TASK_TO_DO = 'ADD_TASK_TO_DO';
export const CHANGE_THE_DESCRIPTION_OF_THE_NEW_TASK = 'CHANGE_THE_DESCRIPTION_OF_THE_NEW_TASK';
export const TASK_HAS_BEEN_ADDED = 'TASK_HAS_BEEN_ADDED';
export const DELETE_TASK = 'DELETE_TASK';
export const TASK_HAS_BEEN_DELETED = 'TASK_HAS_BEEN_DELETED';
export const TASK_HAS_BEEN_UPDATED = 'TASK_HAS_BEEN_UPDATED';
export const UPDATE_TASK = 'UPDATE_TASK';
export const EDIT_TASK = 'EDIT_TASK';
export const CANCEL_EDITING_TASK = 'CANCEL_EDITING_TASK';
export const CHANGE_EDITING_TASK = 'CHANGE_EDITING_TASK';
export const TASKS_REQUEST_FAILED = 'TASKS_REQUEST_FAILED';

/**
 * Get all of the tasks.
 * @returns {{type: string}}
 */
export const getTasks = () => ({
    type: GET_TASKS,
});

/**
 * Tasks request failed.
 * @param {string} errorMessage - Error message
 * @returns {{type: string, payload: {errorMessage: string}}}
 */
export const tasksRequestFailed = errorMessage => ({
    type: TASKS_REQUEST_FAILED,
    payload: {
        errorMessage,
    },
});

/**
 * Tasks have been loaded
 * @param {Array<string>} tasks - Tasks
 * @returns {{type: string, payload: {tasks: Array<string>}}}
 */
export const tasksHaveBeenLoaded = tasks => ({
    type: TASKS_HAVE_BEEN_LOADED,
    payload: {
        tasks,
    },
});


/**
 * Add new task to do
 * @param {string} task - New task
 * @returns {{type: string, payload: {tasks: string}}}
 */
export const addTaskToDo = task => ({
    type: ADD_TASK_TO_DO,
    payload: {
        task,
    },
});


/**
 * Change the description of the new task.
 * @param {string} newTaskDesc - New description of the task
 * @returns {{type: string, payload: {taskId: number, task: string}}}
 */
export const changeTheDescriptionOfTheNewTask = newTaskDesc => ({
    type: CHANGE_THE_DESCRIPTION_OF_THE_NEW_TASK,
    payload: {
        newTaskDesc,
    },
});

/**
 * Task has been added
 * @param {number} taskId - Task id
 * @param {string} task - Task to do
 * @returns {{type: string, payload: {taskId: number, task: string}}}
 */
export const taskHasBeenAdded = (taskId, task) => ({
    type: TASK_HAS_BEEN_ADDED,
    payload: {
        taskId,
        task,
    },
});


/**
 * Delete task
 * @param {number} taskId - Task id
 * @returns {{type: string, payload: {taskId: number}}}
 */
export const deleteTask = taskId => ({
    type: DELETE_TASK,
    payload: {
        taskId,
    },
});

/**
 * Task has been deleted
 * @param {number} taskId - Task id
 * @returns {{type: string, payload: {taskId: number}}}
 */
export const taskHasBeenDeleted = taskId => ({
    type: TASK_HAS_BEEN_DELETED,
    payload: {
        taskId,
    },
});

/**
 * Update task
 * @param {number} taskId - Task id
 * @param {string} newTaskDesc - New task description
 * @returns {{type: string, payload: {taskId: number, newTaskDesc: string}}}
 */
export const updateTask = (taskId, newTaskDesc) => ({
    type: UPDATE_TASK,
    payload: {
        taskId,
        newTaskDesc,
    },
});

/**
 * Task has been updated.
 * @param {number} taskId - Task id
 * @param {string} newTaskDesc - New task description
 * @returns {{type: string, payload: {taskId: number, newTaskDesc: string}}}
 */
export const taskHasBeenUpdated = (taskId, newTaskDesc) => ({
    type: TASK_HAS_BEEN_UPDATED,
    payload: {
        taskId,
        newTaskDesc,
    },
});

/**
 * Edit task
 * @param {number} taskId - Task id
 * @returns {{type: string, payload: {taskId: number}}}
 */
export const editTask = taskId => ({
    type: EDIT_TASK,
    payload: {
        taskId,
    },
});

/**
 * Cancel process of editing task
 * @returns {{type: string}}
 */
export const cancelEditingTask = () => ({
    type: CANCEL_EDITING_TASK,
});

/**
 * Change editing task
 * @param {string} newTaskDesc - New task description
 * @returns {{type: string, payload: {newTaskDesc: string}}}
 */
export const changeEditingTask = newTaskDesc => ({
    type: CHANGE_EDITING_TASK,
    payload: {
        newTaskDesc,
    },
});
