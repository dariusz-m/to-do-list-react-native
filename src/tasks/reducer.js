import { INITIAL_STATE } from '../application-state/initial-state';
import {
    TASKS_HAVE_BEEN_LOADED,
    TASK_HAS_BEEN_ADDED,
    TASK_HAS_BEEN_DELETED,
    EDIT_TASK,
    TASK_HAS_BEEN_UPDATED,
    CANCEL_EDITING_TASK,
    CHANGE_EDITING_TASK,
    CHANGE_THE_DESCRIPTION_OF_THE_NEW_TASK,
    TASKS_REQUEST_FAILED,
} from './actions';

/**
 * Tasks reducer
 * @param {Object} state - To-do sate
 * @param {{type: string, payload: Object}} action - Action
 * @returns {Object}
 */
export default (state = INITIAL_STATE.todo, action) => {
    switch (action.type) {
        case TASKS_HAVE_BEEN_LOADED:
            return { ...state, tasks: action.payload.tasks };
        case TASKS_REQUEST_FAILED:
            return { ...state, errorMessage: action.payload.errorMessage };
        case CHANGE_THE_DESCRIPTION_OF_THE_NEW_TASK:
            return { ...state, newTask: action.payload.newTaskDesc };
        case TASK_HAS_BEEN_ADDED:
            return {
                ...state,
                newTask: INITIAL_STATE.todo.newTask,
                tasks: [...state.tasks, { id: action.payload.taskId, task: action.payload.task }],
            };
        case TASK_HAS_BEEN_DELETED:
            return { ...state, tasks: state.tasks.filter(task => task.id !== action.payload.taskId) };

        case EDIT_TASK:
            return { ...state, editingTask: state.tasks.find(task => task.id === action.payload.taskId) };
        case TASK_HAS_BEEN_UPDATED:
            return {
                ...state,
                editingTask: INITIAL_STATE.todo.editingTask,
                tasks: state.tasks.map(
                    task => (task.id === action.payload.taskId ? { ...task, task: action.payload.newTaskDesc } : task),
                ),
            };
        case CANCEL_EDITING_TASK:
            return { ...state, editingTask: INITIAL_STATE.todo.editingTask };
        case CHANGE_EDITING_TASK:
            return { ...state, editingTask: { ...state.editingTask, task: action.payload.newTaskDesc } };

        default:
            return state;
    }
};
