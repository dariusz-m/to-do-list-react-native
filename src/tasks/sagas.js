import { call, put } from 'redux-saga/effects';

import config from '../config';
import {
    taskHasBeenAdded, taskHasBeenDeleted, taskHasBeenUpdated, tasksHaveBeenLoaded, tasksRequestFailed,
} from './actions';
import { getAccessToken } from '../authentication/access-token';


const DEFAULT_ERROR_MESSAGE = 'Something went wrong, try again.';

/**
 * Get all of the tasks.
 */
export function* getTasks() {
    try {
        const accessToken = yield call(getAccessToken);
        const response = yield call(
            fetch,
            `${config.API_URL}tasks`,
            {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    Authorization: `Bearer ${accessToken}`,
                },
            },
        );
        const data = yield call([response, response.json]);
        if (response.status === 200) {
            yield put(tasksHaveBeenLoaded(data.tasks));
        } else {
            yield put(tasksRequestFailed(data.errorMessage));
        }
    } catch (error) {
        yield put(tasksRequestFailed(DEFAULT_ERROR_MESSAGE));
    }
}

/**
 * Add task to do.
 * @param {{type: string, payload: {task: string}}} action - Action
 */
export function* addTaskToDo(action) {
    try {
        const accessToken = yield call(getAccessToken);
        const response = yield call(
            fetch,
            `${config.API_URL}tasks`,
            {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    Authorization: `Bearer ${accessToken}`,
                },
                body: JSON.stringify({ task: action.payload.task }),
            },
        );
        const data = yield call([response, response.json]);
        if (response.status === 200) {
            yield put(taskHasBeenAdded(data.taskId, action.payload.task));
        } else {
            yield put(tasksRequestFailed(data.errorMessage));
        }
    } catch (error) {
        yield put(tasksRequestFailed(DEFAULT_ERROR_MESSAGE));
    }
}

/**
 * Delete task
 * @param {{type: string, payload: {taskId: number}}} action - Action
 */
export function* deleteTask(action) {
    try {
        const accessToken = yield call(getAccessToken);
        const response = yield call(
            fetch,
            `${config.API_URL}tasks`,
            {
                method: 'DELETE',
                headers: {
                    'Content-Type': 'application/json',
                    Authorization: `Bearer ${accessToken}`,
                },
                body: JSON.stringify({ taskId: action.payload.taskId }),
            },
        );
        if (response.status === 200) {
            yield put(taskHasBeenDeleted(action.payload.taskId));
        } else {
            const data = yield call([response, response.json]);
            yield put(tasksRequestFailed(data.errorMessage));
        }
    } catch (error) {
        yield put(tasksRequestFailed(DEFAULT_ERROR_MESSAGE));
    }
}


/**
 * Update task
 * @param {{type: string, payload: {taskId: number, newTaskDesc: string}}} action - Action
 */
export function* updateTask(action) {
    try {
        const accessToken = yield call(getAccessToken);
        const response = yield call(
            fetch,
            `${config.API_URL}tasks`,
            {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json',
                    Authorization: `Bearer ${accessToken}`,
                },
                body: JSON.stringify({ taskId: action.payload.taskId, newTaskDesc: action.payload.newTaskDesc }),
            },
        );
        if (response.status === 200) {
            yield put(taskHasBeenUpdated(action.payload.taskId, action.payload.newTaskDesc));
        } else {
            const data = yield call([response, response.json]);
            yield put(tasksRequestFailed(data.errorMessage));
        }
    } catch (error) {
        yield put(tasksRequestFailed(DEFAULT_ERROR_MESSAGE));
    }
}
