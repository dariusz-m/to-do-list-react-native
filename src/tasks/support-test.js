// import ReactTestUtils from 'react-dom/test-utils';
// import ReactDOM from 'react-dom';
//
// /**
//  * @param {function} callback - Callback
//  * @param {string|number} firstParameter - First parameter
//  * @param {string|number} secondParameter - Second parameter
//  */
// export const expectedCallbackWasCalledOnceWith = (
//     callback, firstParameter = undefined, secondParameter = undefined,
// ) => {
//     expect(callback.mock.calls.length).toBe(1);
//     expect(callback.mock.calls[0][0]).toEqual(firstParameter);
//     expect(callback.mock.calls[0][1]).toEqual(secondParameter);
// };
//
// /**
//  * @param {React.Component} component - Component
//  * @param {string} buttonId - Button id
//  */
// export const clickButton = (component, buttonId) => {
//     const node = ReactDOM.findDOMNode(component);
//     const button = node.querySelector(`#${buttonId}`);
//     ReactTestUtils.Simulate.click(button);
// };
//
// /**
//  * Change input value
//  * @param {React.Component} component - Component
//  * @param {string} inputId - Input id
//  * @param {string|number} value - Value
//  */
// export const changeInputValue = (component, inputId, value) => {
//     const node = ReactDOM.findDOMNode(component);
//     const field = node.querySelector(`#${inputId}`);
//
//     field.value = value;
//     ReactTestUtils.Simulate.change(field);
// };
