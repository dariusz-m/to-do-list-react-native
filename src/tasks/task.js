import React from 'react';
import PropTypes from 'prop-types';
import {
    Text, View, Button, StyleSheet,
} from 'react-native';

import Input from '../shared-components/input';

/**
 * Task component
 */
export class Task extends React.PureComponent {
    static propTypes = {
        task: PropTypes.shape({
            id: PropTypes.number.isRequired,
            task: PropTypes.string.isRequired,
        }).isRequired,
        deleteTask: PropTypes.func.isRequired,
        editTask: PropTypes.func.isRequired,
        updateTask: PropTypes.func.isRequired,
        cancelEditingTask: PropTypes.func.isRequired,
        changeEditingTask: PropTypes.func.isRequired,
        isEditing: PropTypes.bool.isRequired,
    };

    /**
     * Render
     * @returns {string}
     */
    render() {
        return (
            <View style={styles.container}>
                {
                    this.props.isEditing
                        ? <View>
                            <Input
                                placeholder={''}
                                onChange={value => this.props.changeEditingTask(value)}
                                label={''}
                                autoCorrect={true}
                                secureTextEntry={false}
                                value={this.props.task.task}
                            />
                            <Button onPress={this.props.cancelEditingTask} title={'Cancel'} color={'grey'}/>
                            <Button
                                onPress={() => this.props.updateTask(this.props.task.id, this.props.task.task)}
                                title={'Update'}
                                color={'yellow'}
                            />
                        </View>
                        : <View>
                            <Text>{this.props.task.task}</Text>
                            <Button
                                onPress={() => this.props.editTask(this.props.task.id)}
                                title={'Edit'}
                                color={'blue'}
                            />
                            <Button
                                onPress={() => this.props.deleteTask(this.props.task.id)}
                                title={'Delete'}
                                color={'red'}
                            />
                        </View>
                }
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: '#F5FCFF',
    },
});
