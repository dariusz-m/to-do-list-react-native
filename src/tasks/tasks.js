import React from 'react';
import {
    Text, View, Button, FlatList,
} from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
    addTaskToDo,
    cancelEditingTask,
    changeEditingTask,
    changeTheDescriptionOfTheNewTask,
    deleteTask,
    editTask,
    getTasks,
    updateTask,
} from './actions';
import { Task } from './task';
import Input from '../shared-components/input';

/**
 * Tasks component
 */
export class Tasks extends React.Component {
    static propTypes = {
        tasks: PropTypes.arrayOf(
            PropTypes.shape({
                id: PropTypes.number.isRequired,
                task: PropTypes.string.isRequired,
            }),
        ).isRequired,
        editingTask: PropTypes.shape({
            id: PropTypes.number.isRequired,
            task: PropTypes.string.isRequired,
        }),
        newTask: PropTypes.string.isRequired,
        errorMessage: PropTypes.string,
        actions: PropTypes.shape({
            getTasks: PropTypes.func.isRequired,
            addTaskToDo: PropTypes.func.isRequired,
            deleteTask: PropTypes.func.isRequired,
            editTask: PropTypes.func.isRequired,
            updateTask: PropTypes.func.isRequired,
            cancelEditingTask: PropTypes.func.isRequired,
            changeEditingTask: PropTypes.func.isRequired,
            changeTheDescriptionOfTheNewTask: PropTypes.func.isRequired,
        }).isRequired,
    };

    /**
     * Run when component just have mounted.
     */
    componentDidMount() {
        this.props.actions.getTasks();
    }

    /**
     * Add new task.
     */
    addNewTask = () => {
        this.props.actions.addTaskToDo(this.props.newTask);
    };

    /**
     * Delete task
     * @param {number} taskId - Task id
     */
    deleteTask = (taskId) => {
        this.props.actions.deleteTask(taskId);
    };

    /**
     * Edit task
     * @param {number} taskId - Task id
     */
    editTask = (taskId) => {
        this.props.actions.editTask(taskId);
    };

    /**
     * Update task
     */
    updateTask = () => {
        this.props.actions.updateTask(this.props.editingTask.id, this.props.editingTask.task);
    };

    /**
     * Cancel editing task
     */
    cancelEditingTask = () => {
        this.props.actions.cancelEditingTask();
    };

    /**
     * Change editing task
     * @param {string} newTaskDesc - New task description for editing task.
     */
    changeEditingTask = (newTaskDesc) => {
        this.props.actions.changeEditingTask(newTaskDesc);
    };

    /**
     * Change new task value
     * @param {string} value - Input value.
     */
    onChangeNewTaskInputValue = (value) => {
        this.props.actions.changeTheDescriptionOfTheNewTask(value);
    };

    /**
     * Render
     * @returns {string}
     */
    render() {
        return (
            <View>
                <Text style={{ fontWeight: 'bold' }}>To Do List</Text>
                {this.props.errorMessage
                    ? <Text style={{ color: 'red' }}>{this.props.errorMessage}</Text>
                    : null
                }
                <View style={{ paddingBottom: 30 }}>
                    <Input
                        placeholder={'task...'}
                        onChange={this.onChangeNewTaskInputValue}
                        label={'Add new Task: '}
                        autoCorrect={true}
                        secureTextEntry={false}
                        value={this.props.newTask}
                    />
                    <Button onPress={this.addNewTask} title={'Add'} color={'blue'}/>
                </View>
                <FlatList
                    data={[...this.props.tasks]}
                    keyExtractor={task => task.id.toString()}
                    renderItem={({ item: task }) => {
                        const isTaskEditing = this.props.editingTask ? this.props.editingTask.id === task.id : false;
                        return <Task
                            task={isTaskEditing ? this.props.editingTask : task}
                            deleteTask={this.deleteTask}
                            editTask={this.editTask}
                            isEditing={isTaskEditing}
                            updateTask={this.updateTask}
                            cancelEditingTask={this.cancelEditingTask}
                            changeEditingTask={this.changeEditingTask}
                        />;
                    }}
                />
            </View>
        );
    }
}

/**
 * Map state to props.
 * @param {Object} state - Application state
 * @returns {Object}
 */
const mapStateToProps = state => ({
    tasks: state.todo.tasks,
    editingTask: state.todo.editingTask,
    newTask: state.todo.newTask,
    errorMessage: state.todo.errorMessage,
});

/**
 * Map dispatch to props.
 *
 * @param {Object} dispatch: Redux dispatch
 *
 * @returns {Object}
 */
function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            getTasks,
            addTaskToDo,
            deleteTask,
            editTask,
            updateTask,
            cancelEditingTask,
            changeEditingTask,
            changeTheDescriptionOfTheNewTask,
        }, dispatch),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Tasks);
