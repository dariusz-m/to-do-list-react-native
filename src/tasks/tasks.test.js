import React from 'react';
import renderer from 'react-test-renderer';
import ReactTestUtils from 'react-dom/test-utils';

import { Tasks } from './tasks';
import { changeInputValue, clickButton, expectedCallbackWasCalledOnceWith } from './support-test';

describe('Tasks tests', () => {
    let actions;

    beforeEach(() => {
        actions = {
            getTasks: jest.fn(),
            addTaskToDo: jest.fn(),
            deleteTask: jest.fn(),
            editTask: jest.fn(),
            updateTask: jest.fn(),
            cancelEditingTask: jest.fn(),
            changeEditingTask: jest.fn(),
            changeTheDescriptionOfTheNewTask: jest.fn(),
        };
    });

    afterAll(() => {
        fetch.resetMocks();
    });

    it('Tasks component renders correctly', () => {
        const tasks = [{ id: 1, task: 'Just do it!' }, { id: 2, task: 'Go running!' }];
        const tree = renderer.create(<Tasks tasks={tasks} newTask={''} actions={actions}/>).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Load tasks when component just have mounted', () => {
        ReactTestUtils.renderIntoDocument(
            <Tasks actions={actions} newTask={''} tasks={[]}/>,
        );

        expectedCallbackWasCalledOnceWith(actions.getTasks);
    });

    it('After clicking the button to add new task, callback(addNewTask) is called out once', () => {
        const newTask = 'New task to do!';
        const component = ReactTestUtils.renderIntoDocument(
            <Tasks actions={actions} newTask={newTask} tasks={[]}/>,
        );

        clickButton(component, 'add-new-task-button');

        expectedCallbackWasCalledOnceWith(actions.addTaskToDo, newTask);
    });

    it('Input value is properly updated(onChange value) for new task', () => {
        const newTask = 'Update input value';
        const component = ReactTestUtils.renderIntoDocument(
            <Tasks actions={actions} newTask={''} tasks={[]}/>,
        );

        changeInputValue(component, 'input-new-task', newTask);

        expectedCallbackWasCalledOnceWith(actions.changeTheDescriptionOfTheNewTask, newTask);
    });
});
