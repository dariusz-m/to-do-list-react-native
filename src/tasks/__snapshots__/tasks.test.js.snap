// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks tests Tasks component renders correctly 1`] = `
<View>
  <Text
    style={
      Object {
        "fontWeight": "bold",
      }
    }
  >
    To Do List
  </Text>
  <View
    style={
      Object {
        "paddingBottom": 30,
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Add new Task: 
      </Text>
      <TextInput
        allowFontScaling={true}
        autoCorrect={true}
        onChangeText={[Function]}
        placeholder="task..."
        secureTextEntry={false}
        style={
          Object {
            "borderColor": "black",
            "borderWidth": 0.2,
            "height": 40,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "blue",
              },
            ]
          }
        >
          Add
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": 1,
          "task": "Just do it!",
        },
        Object {
          "id": 2,
          "task": "Go running!",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#F5FCFF",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <View>
            <Text>
              Just do it!
            </Text>
            <View
              accessibilityComponentType="button"
              accessibilityTraits={
                Array [
                  "button",
                ]
              }
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#007AFF",
                        "fontSize": 18,
                        "padding": 8,
                        "textAlign": "center",
                      },
                      Object {
                        "color": "blue",
                      },
                    ]
                  }
                >
                  Edit
                </Text>
              </View>
            </View>
            <View
              accessibilityComponentType="button"
              accessibilityTraits={
                Array [
                  "button",
                ]
              }
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#007AFF",
                        "fontSize": 18,
                        "padding": 8,
                        "textAlign": "center",
                      },
                      Object {
                        "color": "red",
                      },
                    ]
                  }
                >
                  Delete
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#F5FCFF",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <View>
            <Text>
              Go running!
            </Text>
            <View
              accessibilityComponentType="button"
              accessibilityTraits={
                Array [
                  "button",
                ]
              }
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#007AFF",
                        "fontSize": 18,
                        "padding": 8,
                        "textAlign": "center",
                      },
                      Object {
                        "color": "blue",
                      },
                    ]
                  }
                >
                  Edit
                </Text>
              </View>
            </View>
            <View
              accessibilityComponentType="button"
              accessibilityTraits={
                Array [
                  "button",
                ]
              }
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#007AFF",
                        "fontSize": 18,
                        "padding": 8,
                        "textAlign": "center",
                      },
                      Object {
                        "color": "red",
                      },
                    ]
                  }
                >
                  Delete
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
